thunks.go

```diff
@@ -48,118 +44,56 @@ func (rv *readyValue) aPotentialValue() {}
 // potentialValue which guarantees that computation happens at most once).
 type evaluable interface {
 	// fromWhere keeps the information from where the evaluation was requested.
-	getValue(i *interpreter, fromWhere *TraceElement) (value, error)
-}
-
-// thunk holds code and environment in which the code is supposed to be evaluated
-type thunk struct {
-	env  environment
-	body ast.Node
-}
-
-// TODO(sbarzowski) feedback from dcunnin:
-//					makeThunk returning a cachedThunk is weird.
-//					Maybe call thunk 'exprThunk' (or astThunk but then it looks like an AST node).
-//					Then call cachedThunk just thunk?
-//					Or, call this makeCachedExprThunk because that's what it really is.
-func makeThunk(env environment, body ast.Node) *cachedThunk {
-	return makeCachedThunk(&thunk{
-		env:  env,
-		body: body,
-	})
-}
-
-func (t *thunk) getValue(i *interpreter, trace *TraceElement) (value, error) {
-	return i.EvalInCleanEnv(trace, &t.env, t.body, false)
-}
-
-// callThunk represents a concrete, but not yet evaluated call to a function
-type callThunk struct {
-	function evalCallable
-	args     callArguments
-}
-
-func makeCallThunk(ec evalCallable, args callArguments) potentialValue {
-	return makeCachedThunk(&callThunk{function: ec, args: args})
-}
-
-func call(ec evalCallable, arguments ...potentialValue) potentialValue {
-	return makeCallThunk(ec, args(arguments...))
-}
-
-func (th *callThunk) getValue(i *interpreter, trace *TraceElement) (value, error) {
-	evaluator := makeEvaluator(i, trace)
-	err := checkArguments(evaluator, th.args, th.function.Parameters())
-	if err != nil {
-		return nil, err
-	}
-	return th.function.EvalCall(th.args, evaluator)
-}
-
-// deferredThunk allows deferring creation of evaluable until it's actually needed.
-// It's useful for self-recursive structures.
-type deferredThunk func() evaluable
-
-func (th deferredThunk) getValue(i *interpreter, trace *TraceElement) (value, error) {
-	return th().getValue(i, trace)
-}
-
-func makeDeferredThunk(th deferredThunk) potentialValue {
-	return makeCachedThunk(th)
+	getValue(i *interpreter, fromWhere TraceElement) (value, error)
 }
 
 // cachedThunk is a wrapper that caches the value of a potentialValue after
 // the first evaluation.
 // Note: All potentialValues are required to provide the same value every time,
 // so it's only there for efficiency.
-// TODO(sbarzowski) investigate efficiency of various representations
 type cachedThunk struct {
-	pv evaluable
-}
-
-func makeCachedThunk(pv evaluable) *cachedThunk {
-	return &cachedThunk{pv}
+	// The environment is a pointer because it may be a cyclic structure.  A thunk
+	// may refer to itself, so inside `env` there will be a variable bound back to us.
+	env  *environment
+	body ast.Node
+	// If nil, use err.
+	content value
+	// If also nil, content is not cached yet.
+	err error
```

Comment:
If we can do that for values too (rather than forcing a heap allocation for a value) that would be nice.

Reply:
Well, thinking about it now, it could be very problematic - the GC wouldn't like it. Unless we embedded it all the way, but that would get messy I think. The "right" way of doing discriminated unions in Go is via interfaces and type switches - but then we're back at square one.